src/main/kotlin/org/rust/ide/RustfmtWatcher.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide
+
+import com.intellij.AppTopics
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.BaseComponent
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.fileEditor.FileDocumentManagerListener
+import com.intellij.openapi.project.guessProjectForFile
+import com.intellij.util.containers.ContainerUtil
+import org.rust.cargo.project.settings.rustSettings
+import org.rust.cargo.project.settings.toolchain
+import org.rust.openapiext.virtualFile
+
+class RustfmtWatcher : BaseComponent {
+    private val documentsToReformatLater: MutableSet<Document> = ContainerUtil.newConcurrentSet()
+    @Volatile
+    var isEnabled: Boolean = true
+
+    override fun initComponent() {
+        ApplicationManager.getApplication()
+            .messageBus
+            .connect()
+            .subscribe(AppTopics.FILE_DOCUMENT_SYNC, RustfmtListener())
+    }
+
+    fun reformatDocumentLater(document: Document): Boolean {
+        val project = guessProjectForFile(document.virtualFile) ?: return false
+        if (!project.rustSettings.runRustfmtOnSave) return false
+        return documentsToReformatLater.add(document)
+    }
+
+    private inner class RustfmtListener : FileDocumentManagerListener {
+        override fun beforeAllDocumentsSaving() {
+            val documentsToReformat = HashSet(documentsToReformatLater)
+            documentsToReformatLater.clear()
+            for (document in documentsToReformat) {
+                val project = guessProjectForFile(document.virtualFile) ?: continue
+                if (!project.rustSettings.runRustfmtOnSave) continue
+                val rustfmt = project.toolchain?.rustfmt() ?: return
+                rustfmt.reformatDocument(project, document)
+            }
+        }
+
+        override fun beforeDocumentSaving(document: Document) {
+            if (!isEnabled) return
+            val project = guessProjectForFile(document.virtualFile) ?: return
+            if (!project.rustSettings.runRustfmtOnSave) return
+            val rustfmt = project.toolchain?.rustfmt() ?: return
+            rustfmt.reformatDocument(project, document)
```

Comment:
This code is the same as in `beforeAllDocumentsSaving`
Maybe extract it into a separate method?